### PR TITLE
add re routing functionality to route following

### PR DESCRIPTION
**Pull request by saina-ramyar**
# PR Details
## Description

An update should be made to the Route Following Plugin such that if the vehicle is no longer on the shortest route path (due to the operations of another plugin) the Route Following Plugin will generate a lane change or sequence of lane change maneuvers to get the vehicle back onto the shortest route path.

## How Has This Been Tested?
Unit tested

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: route_following_plugin/src/route_following_plugin.cpp [hunk 1]
```diff
@@ -735,4 +749,45 @@ void setManeuverLaneletIds(cav_msgs::Maneuver& mvr, lanelet::Id start_id, lanele
...
+                    lanelet::ConstLanelets interm;
+                    interm.push_back(static_cast<lanelet::ConstLanelet>(target_following_lanelet));
+                    // a new shortest path, via the target_following_lanelet is calculated and used an alternative shortest path
+                    auto new_shortestpath = routing_graph->shortestPathVia(current_lanelet, interm, original_shortestpath.back());
+                    ROS_DEBUG_STREAM("a new shortestpath is generated");
+                    // routeCb is called to update latest_maneuver_plan_
+                    if (new_shortestpath) this->latest_maneuver_plan_ = routeCb(new_shortestpath.get());
+                    break;
+                }
+            }
```
Comment (kjrush) [inline route_following_plugin/src/route_following_plugin.cpp:780]: Do we need an else branch here to handle if no adjacent lanelets are in the shortest path set?

File: route_following_plugin/src/route_following_plugin.cpp [hunk 2]
```diff
@@ -735,4 +749,45 @@ void setManeuverLaneletIds(cav_msgs::Maneuver& mvr, lanelet::Id start_id, lanele
...
+
+    void RouteFollowingPlugin::returnToShortestPath(const lanelet::ConstLanelet &current_lanelet)
+    {
+        auto original_shortestpath = wm_->getRoute()->shortestPath();
+        ROS_DEBUG_STREAM("The vehicle has left the shortest path");
+        auto routing_graph = wm_->getMapRoutingGraph();
+
+        // In order to return to the shortest path, the closest future lanelet on the shortest path needs to be found.
+        // That is the following lanelet of the adjacent lanelet of the current lanelet.
+        auto adjacent_lanelets = routing_graph->besides(current_lanelet);
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply route_following_plugin/src/route_following_plugin.cpp:761]: Here is the definition for that besides(..) call: "Retrieve all reachable left and right lanelets"
I think it means the adjacent lanelets. However, if we are multiple lanelets away, we need to get there one lane-change at a time.

[the remark this answers is not in the document]
Comment (kjrush) [reply route_following_plugin/src/route_following_plugin.cpp:761]: So if that's only returning the directly adjacent lanelets, and we're two lanelets away from the shortest path would the below for loop then fail to find a return path? It looks like we'd loop over the two adjacent lanelets, and if neither if them was on the shortest path we'd fall through and return from this function call without any updates to the plan.

File: route_following_plugin/src/route_following_plugin.cpp (hunk 1 above)
Comment (saina-ramyar) [reply route_following_plugin/src/route_following_plugin.cpp:788]: added

File: route_following_plugin/src/route_following_plugin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply route_following_plugin/src/route_following_plugin.cpp:761]: The goal is that as soon as the vehicle exists the shortest path, this logic willactivate and return the vehicle to the original shortest path. If for some reason we are more than a lanelet away, I added logic that will generate a new shortestpath from the current lanelet of the vehicle.

**Review by kjrush — APPROVED** [review]

File: route_following_plugin/src/route_following_plugin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjrush) [reply route_following_plugin/src/route_following_plugin.cpp:761]: I think having that logic will be good, as my impression was that we'd activate this logic at any point after the vehicle has exited the shortest path and must now return. For example, if the platooning plugin negotiates to platoon in an HOV/platoon lane on the left of a 3-lane highway, but the shortest path is in the right. When we cease platooning at some point we'd need route following to be able to get the vehicle out of the left lane and back into the shortest path.

File: route_following_plugin/src/route_following_plugin.cpp [hunk 3]
```diff
@@ -735,4 +749,56 @@ void setManeuverLaneletIds(cav_msgs::Maneuver& mvr, lanelet::Id start_id, lanele
...
+                    ROS_DEBUG_STREAM("Cannot return to the original shortestpath from adjacent lanes, so a new shortestpath is generated");
+                    // routeCb is called to update latest_maneuver_plan_
+                    if (new_shortestpath) this->latest_maneuver_plan_ = routeCb(new_shortestpath.get());
+                }
+            }
+            
+        }
+        else
+        {
+            ROS_WARN_STREAM("Alternative shortest path cannot be generated");
```
Comment (kjrush) [inline route_following_plugin/src/route_following_plugin.cpp:795]: Should this throw an exception as well? If not, how would the plugin behave if we don't update the shortest path?

Comment (saina-ramyar) [reply route_following_plugin/src/route_following_plugin.cpp:795]: there is a pose callback in route node, that if the vehicles is in an un-routable position will throw an exception. I suggests we let that handle this situation.

**Pull request merged**